### PR TITLE
Add procedure to clean up stack overflow table

### PR DESCRIPTION
**Pull request by tohrnii**
Partly addresses #134. 
Add a procedure to stdlib to remove elements in the stack overflow table to clean up at the end of program completion.
It's kept separate for now and not added to the assembler to include to all programs as suggested by @bobbinth.

### CODE REVIEW

**Comment by tohrnii** [comment]
@bobbinth Should the procedure be kept separate for now as you suggested or should it be added to the assembler? I'd imagine this adds a lot of unnecessary overhead?

File: stdlib/asm/sys.masm [hunk 1]
```diff
@@ -0,0 +1,26 @@
+proc.stack_cleanup_proc.16
+    popw.local.0
+    popw.local.1
+    popw.local.2
+    popw.local.3
+    push.env.sdepth
+    push.16
+    u32gt.unsafe
```
Comment (bobbinth) [inline stdlib/asm/sys.masm:8]: I think we can optimize this a bit. We know that stack depth will never drop below 16, so if stack depth is not 16, we know that it must be greater than 16. This means we should be able to replace "u32gt.unsafe" (which takes 6 VM cycles) with `neq` (which takes 2 VM cycles). We can actually replace `push.16 u32gt.unsafe` with just `neq.16`.

File: stdlib/asm/sys.masm [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    swapw.3
+    loadw.local.2
+    swapw.2
+    loadw.local.1
+    swapw.1
+    loadw.local.0
+end
+
+export.stack_cleanup
+    exec.stack_cleanup_proc
```
Comment (bobbinth) [inline stdlib/asm/sys.masm:25]: We should be able to put the actual code in here - no need to have two separate procedures.

File: stdlib/asm/sys.masm [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    loadw.local.3
+    swapw.3
+    loadw.local.2
+    swapw.2
+    loadw.local.1
+    swapw.1
+    loadw.local.0
+end
+
+export.stack_cleanup
```
Comment (bobbinth) [inline stdlib/asm/sys.masm:24]: Let's rename this to `finalize_stack`. Also, would be great to add brief documentation comments to the function.

File: miden/tests/integration/stdlib/sys.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+use super::build_test;
+use proptest::prelude::*;
+use rand_utils::rand_vector;
+use vm_core::MIN_STACK_DEPTH;
+
+#[test]
+fn stack_cleanup() {
+    let source = "use.std::sys begin push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 push.0 exec.sys::stack_cleanup end";
```
Comment (bobbinth) [inline miden/tests/integration/stdlib/sys.rs:8]: nit: you can probably replace a bunch of `push.0` operations with `repeat.25 push.0 end`.

File: miden/tests/integration/stdlib/sys.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    #[test]
+    fn stack_cleanup_proptest(test_values in prop::collection::vec(any::<u64>(), MIN_STACK_DEPTH), n in 1_usize..100) {
+        let mut push_values = rand_vector::<u64>(n);
+        let mut source_vec = vec!["use.std::sys".to_string(), "begin".to_string()];
+        for value in push_values.iter() {
+            let token = format!("push.{value}");
+            source_vec.push(token);
+        }
+        source_vec.push("exec.sys::stack_cleanup".to_string());
+        source_vec.push("end".to_string());
```
Comment (bobbinth) [inline miden/tests/integration/stdlib/sys.rs:26]: If you use the `repeat` statement inside the script, you can simplify this quite a bit.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks very good! Thank you! I left a few small comments inline.

And yes, let's not add this to the assembler. At leas for now, users can use this procedure manually, if they need to.

File: stdlib/asm/sys.masm (hunk 1 above)
Comment (tohrnii) [reply stdlib/asm/sys.masm:8]: Thank you!

File: stdlib/asm/sys.masm (hunk 2 above)
Comment (tohrnii) [reply stdlib/asm/sys.masm:25]: Thank you. I thought I tried this earlier unsuccessfully but must have been mistaken.

**Comment by tohrnii** [comment]
Thank you so much for the review @bobbinth. Made suggested changes.

File: miden/tests/integration/stdlib/sys.rs (hunk 5 above)
Comment (tohrnii) [reply miden/tests/integration/stdlib/sys.rs]: Sorry, just wanted to clarify, the pushed values are all random (all values are different) so not sure if repeat works? Am I missing something? Although I don't think using random values add any value here.

Comment (bobbinth) [reply miden/tests/integration/stdlib/sys.rs]: Ah - i see! Yes, it's fine as is then.

**Review by bobbinth — APPROVED** [review]
Looks great! Thank you again.

**Pull request merged**